tests/leaderelection_test.go

```diff
@@ -0,0 +1,142 @@
+
+	pollingInterval = 2 * time.Second
+	timeout         = 90 * time.Second
+)
+
+var (
+	logCheckLeaderRegEx = regexp.MustCompile("Attempting to acquire leader lease")
+	logIsLeaderRegex    = regexp.MustCompile("Successfully acquired leadership lease")
+)
+
+func checkLogForRegEx(regEx *regexp.Regexp, log string) bool {
+	matches := regEx.FindAllStringIndex(log, -1)
+	return len(matches) == 1
+}
+
+var _ = Describe("Leader election tests", func() {
+	var err error
+	var deployment *appsv1.Deployment
+	var leaderPodName string
+
+	f := framework.NewFrameworkOrDie("leaderelection-test")
+
+	getDeployment := func() *appsv1.Deployment {
+		deployment, err := f.K8sClient.AppsV1().Deployments(f.CdiInstallNs).Get(cdiDeploymentName, metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		return deployment
+	}
+
+	getPods := func() *v1.PodList {
+		pods, err := f.K8sClient.CoreV1().Pods(f.CdiInstallNs).List(metav1.ListOptions{LabelSelector: "app=containerized-data-importer"})
+		Expect(err).ToNot(HaveOccurred())
+		return pods
+	}
+
+	getLog := func(name string) string {
+		log, err := tests.RunKubectlCommand(f, "logs", name, "-n", f.CdiInstallNs)
+		Expect(err).ToNot(HaveOccurred())
+		return log
+	}
+
+	AfterEach(func() {
+		deployment = getDeployment()
+		if *deployment.Spec.Replicas > 1 {
+			n := int32(1)
+			deployment.Spec.Replicas = &n
+			_, err = f.K8sClient.AppsV1().Deployments(f.CdiInstallNs).Update(deployment)
+			Expect(err).ToNot(HaveOccurred())
+
+			Eventually(func() []v1.Pod {
+				return getPods().Items
+			}, timeout, pollingInterval).Should(HaveLen(1))
+		}
+	})
+
+	It("Should be one active CDI controller", func() {
+		deployment = getDeployment()
+		Expect(*deployment.Spec.Replicas).Should(Equal(int32(1)))
+
+		pods := getPods()
+		Expect(pods.Items).Should(HaveLen(1))
+
+		leaderPodName = pods.Items[0].Name
+
+		log := getLog(leaderPodName)
+		Expect(checkLogForRegEx(logIsLeaderRegex, log)).To(BeTrue())
+	})
+
+	It("Scale controller deployment", func() {
+		numReplicas := int32(2)
+		deployment.Spec.Replicas = &numReplicas
+
+		deployment, err = f.K8sClient.AppsV1().Deployments(f.CdiInstallNs).Update(deployment)
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(func() bool {
+			pods := getPods()
+			if len(pods.Items) != 2 {
+				return false
+			}
+			for _, pod := range pods.Items {
+				if pod.Status.Phase != "Running" {
+					return false
+				}
+			}
+			return true
+
+		}, timeout, pollingInterval).Should(BeTrue())
+
+		newPodName := ""
+		pods := getPods().Items
+		Expect(pods).To(HaveLen(2))
+		for _, pod := range getPods().Items {
+			if pod.Name != leaderPodName {
+				newPodName = pod.Name
+				break
+			}
+		}
+		Expect(newPodName).ShouldNot(Equal(""))
+
+		Eventually(func() bool {
+			log := getLog(newPodName)
+			return checkLogForRegEx(logCheckLeaderRegEx, log)
+		}, timeout, pollingInterval).Should(BeTrue())
+
+		// have to prove new pod won't become leader in period longer than lease duration
```

Comment:
Can we shorten the lease duration for tests only so we don't have such a long sleep?

Reply:
20 seconds might be a little short actually. Especially in a highly oversubscribed environment like STDCI, where simple operations can take 10 seconds or longer.

Reply:
@aglitke changing the lease duration for the functional test would basically require that lease duration is a configurable parameter (command line arg or environment variable) and the functional test can edit the the deployment and restart the cdi controller with different params.  I don't think this in in the spirit of functional tests.  I already hate that we scale out the cdi-deployment for this.  It's risky.  Could taint the test environment somehow.  

I'm really on the fence of whether this test should exist at all.  The leader election functionality is already covered by tests in kubernetes client-go.  A reviewer really just needs to verify that the appropriate function is called when we become leader and that we shoot ourselves in the head if we are no longer leader.  The rest is basically boilerplate crap.

```
Callbacks: leaderelection.LeaderCallbacks{
	OnStartedLeading: func(_ <-chan struct{}) {
		glog.Info("Successfully acquired leadership lease")
		leaderFunc()
	},
	OnStoppedLeading: func() {
		glog.Fatal("NO LONGER LEADER, EXITING")

	},
},
```

Also, when we switch to controller-runtime this code will be deleted.

Reply:
@awels, what would you suggest for a timeout?